File: backend/src/main/java/com/smvita/computerseekho/service/CounselorAssignmentService.java
package com.smvita.computerseekho.service;

import com.smvita.computerseekho.entity.Inquiry;
import com.smvita.computerseekho.entity.Staff;
import com.smvita.computerseekho.repository.InquiryRepository;
import com.smvita.computerseekho.repository.StaffRepository;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.Set;

/**
 * KB §9.6: counselor auto-assignment = least-loaded round robin.
 *
 * "Least-loaded" is measured as the number of enquiries the counselor is
 * still actively working (New or In-Followup) — closed and converted
 * enquiries are finished work and shouldn't count against them. Ties are
 * broken round-robin style, by whoever was handed an enquiry least
 * recently, so an idle team doesn't funnel every new lead to the counselor
 * with the lowest staff_id.
 *
 * Pulled out of InquiryService into its own bean because it's the single
 * highest-value piece of business logic on Day 3 and the Day 5 plan calls
 * for Mockito tests on exactly this — far easier to test in isolation.
 */
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class CounselorAssignmentService {

    private static final Logger log = LoggerFactory.getLogger(CounselorAssignmentService.class);

    /** An enquiry still on someone's plate. */
    public static final Set<Inquiry.Status> OPEN_STATUSES =
            Set.of(Inquiry.Status.New, Inquiry.Status.InFollowup);

    private final StaffRepository staffRepository;
    private final InquiryRepository inquiryRepository;

    /**
     * Returns the counselor who should take the next enquiry, or empty if
     * there are no active Counselors at all. Callers treat empty as
     * "unassigned" rather than an error: an enquiry must never be lost just
     * because nobody is set up to receive it — it stays visible with no
     * counselor until an admin assigns one.
     */
    public Optional<Staff> nextCounselor() {
        List<Staff> counselors = staffRepository.findByRoleAndIsActiveTrue(Staff.Role.Counselor);
        if (counselors.isEmpty()) {
            log.warn("No active Counselor to auto-assign this enquiry to — leaving it unassigned");
            return Optional.empty();
        }

        // Built as a named local rather than inline: starting a comparator
        // chain with Comparator.comparingLong(this::openLoad) leaves the
        // element type in a receiver position, where javac can't infer it.
        Comparator<Staff> leastLoadedThenLongestWaiting =
                Comparator.<Staff>comparingLong(this::openLoad)
                        .thenComparingInt(this::lastAssignedInquiryId)
                        .thenComparingInt(Staff::getStaffId);

        Optional<Staff> chosen = counselors.stream().min(leastLoadedThenLongestWaiting);

        chosen.ifPresent(s -> log.info("Auto-assigned enquiry to counselor '{}' (open load {})",
                s.getName(), openLoad(s)));
        return chosen;
    }

    private long openLoad(Staff counselor) {
        return inquiryRepository.countByStaff_StaffIdAndStatusIn(counselor.getStaffId(), OPEN_STATUSES);
    }

    /**
     * Highest inquiry_id this counselor has ever been given; 0 if never.
     * Lower value == waited longer for their turn, which is what makes the
     * tie-break a rotation rather than a fixed ordering.
     */
    private int lastAssignedInquiryId(Staff counselor) {
        Inquiry latest = inquiryRepository.findTopByStaff_StaffIdOrderByInquiryIdDesc(counselor.getStaffId());
        return latest == null ? 0 : latest.getInquiryId();
    }
}

File: dotnet_backend/backend/Service/CounselorAssignmentService.cs
using ComputerSeekho.Models;
using ComputerSeekho.Repository;

namespace ComputerSeekho.Service;

/// <summary>
/// Least-loaded round robin.
///
/// Load counts only *open* enquiries — New and In-Followup. Counting
/// everything would mean a counsellor who has closed a hundred leads looks
/// permanently busiest and stops receiving work, which is the opposite of
/// what load balancing is for.
///
/// Ties are broken by whoever was assigned least recently, so an idle team
/// rotates instead of every enquiry landing on the lowest staff id.
/// </summary>
public class CounselorAssignmentService : ICounselorAssignmentService
{
    /// <summary>An enquiry still being worked. Closed ones carry no load.</summary>
    private static readonly InquiryStatus[] OpenStatuses =
        new[] { InquiryStatus.New, InquiryStatus.InFollowup };

    private readonly IGenericRepository<Staff> _staff;
    private readonly IGenericRepository<Inquiry> _inquiries;
    private readonly ILogger<CounselorAssignmentService> _logger;

    public CounselorAssignmentService(
        IGenericRepository<Staff> staff,
        IGenericRepository<Inquiry> inquiries,
        ILogger<CounselorAssignmentService> logger)
    {
        _staff = staff;
        _inquiries = inquiries;
        _logger = logger;
    }

    public async Task<Staff?> NextCounselorAsync(CancellationToken ct = default)
    {
        var counselors = (await _staff.FindAsync(
            s => s.Role == StaffRole.Counselor && s.IsActive, ct)).ToList();

        if (counselors.Count == 0)
        {
            // Not an error. The enquiry is still captured; the enquiry list's
            // Schedule link is how it gets picked up later.
            _logger.LogWarning("No active counsellor to assign — enquiry will be unassigned");
            return null;
        }

        // One read rather than a count query per counsellor. The open set is
        // small by definition, and N queries for N counsellors is a poor
        // trade for a list that fits in memory.
        var openInquiries = (await _inquiries.FindAsync(
            i => (i.Status == InquiryStatus.New || i.Status == InquiryStatus.InFollowup)
                 && i.StaffId != null, ct)).ToList();

        var chosen = counselors
            .OrderBy(c => openInquiries.Count(i => i.StaffId == c.StaffId))
            // Least recently assigned wins a tie. Using the highest inquiry
            // id they hold as a proxy for recency: a counsellor with none at
            // all sorts first, which is what should happen.
            .ThenBy(c => openInquiries
                .Where(i => i.StaffId == c.StaffId)
                .Select(i => (int?)i.InquiryId)
                .Max() ?? 0)
            .First();

        _logger.LogInformation("Assigning enquiry to {Counselor} (staff {StaffId})",
            chosen.Name, chosen.StaffId);

        return chosen;
    }
}
